apm-sniffer/apm-sdk-plugin/undertow-plugins/undertow-2.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/undertow/v2x/util/TraceContextUtils.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package org.apache.skywalking.apm.plugin.undertow.v2x.util;
+
+import io.undertow.server.HttpServerExchange;
+import io.undertow.util.HeaderMap;
+import org.apache.skywalking.apm.agent.core.context.CarrierItem;
+import org.apache.skywalking.apm.agent.core.context.ContextCarrier;
+import org.apache.skywalking.apm.agent.core.context.ContextManager;
+import org.apache.skywalking.apm.agent.core.context.tag.Tags;
+import org.apache.skywalking.apm.agent.core.context.trace.AbstractSpan;
+import org.apache.skywalking.apm.agent.core.context.trace.SpanLayer;
+import org.apache.skywalking.apm.network.trace.component.ComponentsDefine;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * @author AI
+ * 2019-07-26
+ */
+public class TraceContextUtils {
+
+    private static final AtomicBoolean IS_IN_ROUTING_HANDLER_TRACE = new AtomicBoolean(false);
```

Comment:
This is a static final field, what happens in the concurrency case?

Reply:
This is a private field, When the agent was bind `RoutingHandlerInterceptor` to method `io.undertow.server.RoutingHandler.add` the field will be set to `true`.

Why need the flag(`IS_IN_ROUTING_HANDLER_TRACE`) ?
The `ExecuteRootHandlerInterceptor`(it is a static method plugin) will be bind to method `io.undertow.server.Connectors.executeRootHandler` also in agent started. When undertow handling user request. `io.undertow.server.Connectors.executeRootHandler` and `io.undertow.server.HttpHandler.handleRequest` will be invoked both and request will be traced twice (it will create two EntrySpan).

We need a flag avoid a request traced twice. So when the agent was binded `RoutingHandlerInterceptor` it not trace in `ExecuteRootHandlerInterceptor` no more.

Reply:
> When undertow handling user request. io.undertow.server.Connectors.executeRootHandler and io.undertow.server.HttpHandler.handleRequest will be invoked both and request will be traced twice

Could we make the binding happens once? Or this keeps in uncertain status.

> it will create two EntrySpan
These two in different threads? If in the same thread, these two should be merged automatically.

Reply:
> Could we make the binding happens once? Or this keeps in uncertain status.

No, we cannot.

When undertow using `RoutingHandler` it will binding to `RoutingHandlerInterceptor`
```
RoutingHandler handler = new RoutingHandler();
// RoutingHandlerInterceptor's cutpoint is add method
handler.add(Methods.GET, "/projects/{projectId}", exchange -> {
    exchange.getResponseHeaders().put(Headers.CONTENT_TYPE, "text/plain");
    exchange.getResponseSender().send("Hello World");
 });
Undertow server = Undertow.builder().setHandler(handler)...
```
or using `HttpHandler` it will binding to `ExecuteRootHandlerInterceptor`
```
io.undertow.server.Connectors.executeRootHandler
 Undertow.builder().setHandler(new HttpHandler() {
    @Override
    public void handleRequest(HttpServerExchange httpServerExchange) throws Exception {
        //
    }
});
```
We cannot detected user how to use `undertow`. So we need keep `ExecuteRootHandlerInterceptor` and `RoutingHandlerInterceptor` both.

> These two in different threads? 

Maybe, `HttpServerExchange` can dispatched to other `Executor`.

> If in the same thread, these two should be merged automatically.

EntrySpan 1: /projects/1
EntrySpan 2: /projects/{projectId} 

it can merged automatically?

Reply:
If EntrySpan2 happens after EntrySpan1, and both of them in the same thread, then yes. it will merge automatically. You could try.

Reply:
Thanks

Reply:
What is the progress here? Is merge working? Even not, there is still an issue. Think in this way, what happens in user set up two `Undertow` servers w/ different handler? The value will be changed in the wrong way.